dev/ppc64le/src/intt_ppc.S

```diff
@@ -0,0 +1,672 @@
+	vmhraddshs 30, 30, V_NMKQ, 29
+
+	vsrah \_vo0, 15, 4		# >> 1
+	vsrah \_vo1, 20, 4		# >> 1
+	vsrah \_vo2, 25, 4		# >> 1
+	vsrah \_vo3, 30, 4		# >> 1
+.endm
+
+.macro Set_mont_consts
+	xxlor	32+5, 0, 0	# V_NMKQ
+	xxlor	32+2, 2, 2	# V_QINV
+	xxlor	32+3, 3, 3	# 0
+	xxlor	32+4, 4, 4	# 1
+.endm
+
+.macro Load_next_4zetas
+	lxv	32+V_Z0, 0(14)
+	lxv	32+V_Z1, 16(14)
+	lxv	32+V_Z2, 32(14)
+	lxv	32+V_Z3, 48(14)
+	addi	14, 14, 64
+.endm
+
+.macro Write_B4C _vs0 _vs1 _vs2 _vs3
+	stxvx	\_vs0, 3, 9
+	stxvx	\_vs1, 3, 16
+	stxvx	\_vs2, 3, 18
+	stxvx	\_vs3, 3, 20
+.endm
+
+.macro Write_M4C _vs0 _vs1 _vs2 _vs3
+	stxvx	\_vs0, 3, 10
+	stxvx	\_vs1, 3, 17
+	stxvx	\_vs2, 3, 19
+	stxvx	\_vs3, 3, 21
+.endm
+
+.macro Reload_4coeffs
+	lxv	32+25, 0(3)
+	lxv	32+26, 16(3)
+	lxv	32+30, 32(3)
+	lxv	32+31, 48(3)
+	addi	3, 3, 64
+.endm
+
+.macro MWrite_8X _vs0 _vs1 _vs2 _vs3 _vs4 _vs5 _vs6 _vs7
+	stxv	\_vs0, -128(3)
+	stxv	\_vs1, -112(3)
+	stxv	\_vs2, -96(3)
+	stxv	\_vs3, -80(3)
+	stxv	\_vs4, -64(3)
+	stxv	\_vs5, -48(3)
+	stxv	\_vs6, -32(3)
+	stxv	\_vs7, -16(3)
+.endm
+
+.macro Write_Len2_4C _vs0 _vs1 _vs2 _vs3
+        xxmrglw 32+12, \_vs0, 10
+        xxmrghw 32+11, \_vs0, 10
+        xxpermdi 10, 32+12, 32+11, 3
+        xxmrglw 32+16, \_vs1, 11
+        xxmrghw 32+15, \_vs1, 11
+        xxpermdi 11, 32+16, 32+15, 3
+        xxmrglw 32+12, \_vs2, 12
+        xxmrghw 32+11, \_vs2, 12
+        xxpermdi 12, 32+12, 32+11, 3
+        xxmrglw 32+16, \_vs3, 13
+        xxmrghw 32+15, \_vs3, 13
+        xxpermdi 13, 32+16, 32+15, 3
+        stxvd2x   10, 3, 9
+        stxvd2x   11, 3, 16
+        stxvd2x   12, 3, 18
+        stxvd2x   13, 3, 20
+.endm
+
+.macro Write_Len4_4C _vs0 _vs1 _vs2 _vs3
+	xxpermdi 10, 10, \_vs0, 3
+	xxpermdi 11, 11, \_vs1, 3
+	xxpermdi 12, 12, \_vs2, 3
+	xxpermdi 13, 13, \_vs3, 3
+        stxvd2x   10, 3, 9
+        stxvd2x   11, 3, 16
+        stxvd2x   12, 3, 18
+        stxvd2x   13, 3, 20
+.endm
+
+# intt
+# t = r[j];
+# r[j] = barrett_reduce(t + r[j + len]);
+# r[j + len] = r[j + len] - t;
+# r[j + len] = fqmul(zeta, r[j + len]);
+
+#
+# mlk_intt_ppc(r)
+#
+.global MLK_ASM_NAMESPACE(intt_ppc)
+.align 4
+MLK_ASM_FN_SYMBOL(intt_ppc)
+
+	stdu	1, -352(1)
+	mflr	0
+	std	14, 56(1)
+	std	15, 64(1)
+	std	16, 72(1)
+	std	17, 80(1)
+	std	18, 88(1)
+	std	19, 96(1)
+	std	20, 104(1)
+	std	21, 112(1)
+	stxv	32+20, 128(1)
+	stxv	32+21, 144(1)
+	stxv	32+22, 160(1)
+	stxv	32+23, 176(1)
+	stxv	32+24, 192(1)
+	stxv	32+25, 208(1)
+	stxv	32+26, 224(1)
+	stxv	32+27, 240(1)
+	stxv	32+28, 256(1)
+	stxv	32+29, 272(1)
+	stxv	32+30, 288(1)
+	stxv	32+31, 304(1)
+
+	# init vectors and constants
+	# Setup for Montgomery reduce
+	lxv	0, 0(4)
+
+	lxv	32+V_QINV, QINV_OFFSET(4)       # QINV
+	xxlxor	32+3, 32+3, 32+3
+	vspltish 4, 1
+	xxlor	2, 32+2, 32+2			# QINV
+	xxlor	3, 32+3, 32+3			# 0
+	xxlor	4, 32+4, 32+4			# 1
+
+	# Setup for Barrett reduce
+	lxv	6, Q_OFFSET(4)			# V_MKQ
+	lxv	32+V20159, C20159_OFFSET(4)	# V20159
+	lxv	7, 0(4)	# V_25
+
+	#xxspltiw 8, 26		# for power9 and above
```

Comment:
What else needs changing on power9? Which code would be commented in, which would be commented out?